src/lib/mathlib/math/filter/second_order_reference_model.hpp

```diff
@@ -0,0 +1,197 @@
+	 */
+	bool setParameters(const float natural_freq, const float damping_ratio)
+	{
+		if (natural_freq < FLT_EPSILON || damping_ratio < FLT_EPSILON) {
+
+			// Deadzone the resulting constants (will result in zero filter acceleration)
+			spring_constant_ = 0.0f;
+			damping_coefficient_ = 0.0f;
+
+			// Zero natural frequency means our time step is irrelevant
+			max_time_step_ = INFINITY;
+
+			// Fail
+			return false;
+		}
+
+		// Note these are "effective" coefficients, as mass coefficient is considered baked in
+		spring_constant_ = natural_freq * natural_freq;
+		damping_coefficient_ = 2.0f * damping_ratio * natural_freq;
+
+		// Based on *conservative nyquist frequency via SAMPLE_RATE_MULTIPLIER
+		max_time_step_ = 2.0f * M_PI_F / (natural_freq * SAMPLE_RATE_MULTIPLIER);
+
+		return true;
+	}
+
+	/**
+	 * @return System state [units]
+	 */
+	const T &getState() const { return filter_state_; }
+
+	/**
+	 * @return System rate [units/s]
+	 */
+	const T &getRate() const { return filter_rate_; }
+
+	/**
+	 * @return System acceleration [units/s^2]
+	 */
+	const T &getAccel() const { return filter_accel_; }
+
+	/**
+	 * Update the system states
+	 *
+	 * Units of rate_sample must correspond to the derivative of state_sample units.
+	 * There is currently no handling of discrete time behavior w.r.t. continuous time
+	 * system parameters. Sampling time should be sufficiently fast.
+	 *
+	 * @param[in] time_step Time since last sample [s]
+	 * @param[in] state_sample New state sample [units]
+	 * @param[in] rate_sample New rate sample, if provided, otherwise defaults to zero(s) [units/s]
+	 */
+	void update(const float time_step, const T &state_sample, const T &rate_sample = T())
+	{
+		if (time_step > max_time_step_) {
+			// time step is too large, reset the filter
+			reset(state_sample, rate_sample);
+			return;
+		}
+
+		if (time_step < 0.0f) {
+			// erroneous input, dont update
+			return;
+		}
+
+		filter_state_ = integrate(filter_state_, filter_rate_, time_step);
+		filter_rate_ = integrate(filter_rate_, filter_accel_, time_step);
+
+		T state_error = state_sample - filter_state_;
+		T rate_error = rate_sample - filter_rate_;
+		filter_accel_ = state_error * spring_constant_ + rate_error * damping_coefficient_;
+	}
+
+	/**
+	 * Reset the system states
+	 *
+	 * @param[in] state Initial state [units]
+	 * @param[in] rate Initial rate, if provided, otherwise defaults to zero(s) [units/s]
+	 */
+	void reset(const T &state, const T &rate = T())
+	{
+		filter_state_ = state;
+		filter_rate_ = rate;
+		filter_accel_ = T();
+	}
+
+protected:
+
+	// A conservative multiplier on sample frequency to bound the maximum time step
+	static constexpr float SAMPLE_RATE_MULTIPLIER = 10.0f;
+
+	// (effective, no mass) Spring constant for second order system [s^-2]
+	float spring_constant_{0.0f};
+
+	// (effective, no mass) Damping coefficient for second order system [s^-1]
+	float damping_coefficient_{0.0f};
+
+	T filter_state_{}; // [units]
```

Comment:
```suggestion
	T _filter_state{}; // [units]
```
Not really important, but we're usually **prefixing** with underscore for member variables

Reply:
ill wait until our code style discussion/decision finalizes to change this :)